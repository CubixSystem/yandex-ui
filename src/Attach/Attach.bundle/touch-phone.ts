import { withRegistry, Registry } from '@bem-react/di';
import { compose, composeU } from '@bem-react/core';

// base
import { Attach as AttachTouchPhone, cnAttach } from '../Attach@touch-phone';
// _size
import { withSizeL } from '../_size/Attach_size_l';
import { withSizeM } from '../_size/Attach_size_m';
import { withSizeS } from '../_size/Attach_size_s';

// Registry dependencies â€” Button
import { Button as ButtonDesktop } from '../../Button/Button@touch-phone';
import { withSizeL as withButtonSizeL } from '../../Button/_size/Button_size_l';
import { withSizeM as withButtonSizeM } from '../../Button/_size/Button_size_m';
import { withSizeS as withButtonSizeS } from '../../Button/_size/Button_size_s';
import { withViewDefault as withButtonViewDefault } from '../../Button/_view/Button_view_default';
import { withThemeNormal as withButtonThemeNormal } from '../../Button/_theme/Button_theme_normal';

const registry = new Registry({ id: cnAttach() });

const Button = compose(
    withButtonThemeNormal,
    withButtonViewDefault,
    composeU(withButtonSizeL, withButtonSizeM, withButtonSizeS),
)(ButtonDesktop);

registry.set('Button', Button);

export * from '../Attach@touch-phone';

export const Attach = compose(
    composeU(withSizeL, withSizeM, withSizeS),
    withRegistry(registry),
)(AttachTouchPhone);
