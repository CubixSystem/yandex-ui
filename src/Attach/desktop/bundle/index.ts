export * from '../../Attach.bundle/desktop';
