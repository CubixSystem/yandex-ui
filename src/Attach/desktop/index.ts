export * from '../Attach@desktop';
export * from '../_size/Attach_size_l';
export * from '../_size/Attach_size_m';
export * from '../_size/Attach_size_s';
