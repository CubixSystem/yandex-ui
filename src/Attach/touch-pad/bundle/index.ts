export * from '../../Attach.bundle/touch-pad';
