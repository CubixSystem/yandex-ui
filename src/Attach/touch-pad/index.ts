export * from '../Attach@touch-pad';
export * from '../_size/Attach_size_l';
export * from '../_size/Attach_size_m';
export * from '../_size/Attach_size_s';
