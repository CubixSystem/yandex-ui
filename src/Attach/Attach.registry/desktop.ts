import { Registry } from '@bem-react/di';

import { cnAttach } from '../Attach.const';
import { Holder } from '../Holder/Attach-Holder';
import { Control } from '../Control/Attach-Control';

export const attachRegistry = new Registry({ id: cnAttach() });

attachRegistry
    .set('Holder', Holder)
    .set('Control', Control)
    .set('Button', () => {
        // TODO: Использовать error api из новой версии react.
        throw new Error(`Для работы "Attach" требуется установить компонент "Button" в реестр "${cnAttach()}".`);
    });
