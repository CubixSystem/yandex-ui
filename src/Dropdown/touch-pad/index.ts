import { ExtractProps } from '@bem-react/core';
import { ComponentType } from 'react';
import { withDropdown, DropdownProps } from '../Dropdown';
import { Popup } from '../../Popup/Popup.bundle/touch-pad';

export const Dropdown = withDropdown<ExtractProps<typeof Popup>>(Popup) as ComponentType<DropdownProps>;
export * from '../Dropdown';
