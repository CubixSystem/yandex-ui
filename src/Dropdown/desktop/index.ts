import { ExtractProps } from '@bem-react/core';
import { withDropdown } from '../Dropdown';
import { Popup } from '../../Popup/Popup.bundle/desktop';

export const Dropdown = withDropdown<ExtractProps<typeof Popup>>(Popup);

export * from '../Dropdown';
