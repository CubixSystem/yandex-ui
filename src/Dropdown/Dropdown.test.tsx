import React from 'react';
import { mount, render, ReactWrapper } from 'enzyme';
import { Dropdown as DropdownDesktop } from './desktop';
import { Dropdown as DropdownTouchPhone } from './touch-phone';
import { Dropdown as DropdownTouchPad } from './touch-pad';
import { Popup } from '../Popup/Popup';
import { Button } from '../Button/desktop/bundle';

const platforms = [['desktop', DropdownDesktop], ['touch-phone', DropdownTouchPhone], ['touch-pad', DropdownTouchPad]];

describe.each<any>(platforms)('Dropdown@s', (_platform, Dropdown) => {
    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        expect(
            render(
                <Dropdown content={<div>Test</div>}>
                    <Button />
                </Dropdown>,
            ),
        ).toMatchSnapshot();
    });
});

describe.each<any>(platforms.slice(0, 1))('Dropdown@%s', (_platform, Dropdown) => {
    let wrapper: ReactWrapper | null;

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });

    test('должен рендерить кнопку', () => {
        wrapper = mount(
            <Dropdown content={<div>Text</div>}>
                <Button />
            </Dropdown>,
        );
        expect(wrapper.find(Button)).toHaveLength(1);
    });

    test('должен рендерить попап', () => {
        wrapper = mount(
            <Dropdown content={<div>Text</div>}>
                <Button />
            </Dropdown>,
        );
        expect(wrapper.find(Popup)).toHaveLength(1);
    });

    test('должен рендерить видимый попап', () => {
        wrapper = mount(
            <Dropdown visible content={<div>Text</div>}>
                <Button />
            </Dropdown>,
        );
        expect(wrapper.find(Popup).prop('visible')).toBeTruthy();
    });

    test('должен вызывать `onVisibleChange` c правильными аргументами по клику', () => {
        const visibleChangeHandler = jest.fn();
        wrapper = mount(
            <Dropdown onVisibleChange={visibleChangeHandler} trigger="click" content={<div>Text</div>}>
                <Button />
            </Dropdown>,
        );

        wrapper.find(Button).simulate('click');
        expect(visibleChangeHandler).toHaveBeenCalledTimes(1);
        wrapper.find(Button).simulate('click');
        expect(visibleChangeHandler).toHaveBeenCalledTimes(2);
        expect(visibleChangeHandler).toHaveBeenNthCalledWith(1, true);
        expect(visibleChangeHandler).toHaveBeenNthCalledWith(2, false);

        expect(visibleChangeHandler).toHaveBeenNthCalledWith(1, true);
        expect(visibleChangeHandler).toHaveBeenNthCalledWith(2, false);
    });

    test('должен вызывать `onVisibleChange` c правильными аргументами по наведению', () => {
        const visibleChangeHandler = jest.fn();
        jest.useFakeTimers();

        wrapper = mount(
            <Dropdown onVisibleChange={visibleChangeHandler} content={<div>Text</div>}>
                <Button />
            </Dropdown>,
        );

        wrapper.find(Button).simulate('mouseenter');
        jest.runOnlyPendingTimers();
        expect(visibleChangeHandler).toHaveBeenCalledTimes(1);
        wrapper.find(Button).simulate('mouseleave');
        jest.runOnlyPendingTimers();
        expect(visibleChangeHandler).toHaveBeenCalledTimes(2);

        expect(visibleChangeHandler).toHaveBeenNthCalledWith(1, true);
        expect(visibleChangeHandler).toHaveBeenNthCalledWith(2, false);
    });

    test('должен вызывать `onVisibleChange` c правильными аргументами по фокусу', () => {
        const visibleChangeHandler = jest.fn();
        jest.useFakeTimers();

        wrapper = mount(
            <Dropdown onVisibleChange={visibleChangeHandler} trigger="focus" content={<div>Text</div>}>
                <Button view="default" size="s">
                    Dropdown
                </Button>
            </Dropdown>,
        );

        wrapper.find(Button).simulate('focus');
        jest.runOnlyPendingTimers();
        expect(visibleChangeHandler).toHaveBeenCalledTimes(1);
        wrapper.find(Button).simulate('blur');
        jest.runOnlyPendingTimers();
        expect(visibleChangeHandler).toHaveBeenCalledTimes(2);

        expect(visibleChangeHandler).toHaveBeenNthCalledWith(1, true);
        expect(visibleChangeHandler).toHaveBeenNthCalledWith(2, false);
    });

    test('должен вызывать колбэк переданный в кнопке', () => {
        const handler = jest.fn();
        wrapper = mount(
            <Dropdown trigger={['click', 'hover']} content={<div>Text</div>}>
                <Button onMouseLeave={handler} onClick={handler} view="default" size="s">
                    Dropdown
                </Button>
            </Dropdown>,
        );

        wrapper.find(Button).simulate('click');
        expect(handler).toHaveBeenCalledTimes(1);
        wrapper.find(Button).simulate('mouseleave');
        expect(handler).toHaveBeenCalledTimes(2);
    });
});
