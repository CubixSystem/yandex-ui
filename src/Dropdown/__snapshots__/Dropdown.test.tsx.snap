// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown@s должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="Dropdown"
>
  <button
    autocomplete="off"
    class="Button2"
    type="button"
  />
</span>
`;

exports[`Dropdown@s должен вернуть полный шаблон компонента (snapshot) 2`] = `
<span
  class="Dropdown"
>
  <button
    autocomplete="off"
    class="Button2"
    type="button"
  />
</span>
`;

exports[`Dropdown@s должен вернуть полный шаблон компонента (snapshot) 3`] = `
<span
  class="Dropdown"
>
  <button
    autocomplete="off"
    class="Button2"
    type="button"
  />
</span>
`;
